kubernetes/scheduler/controller.go

```diff
@@ -0,0 +1,668 @@
+	//      // Get logs
+	//      logs, err := c.libclient.GetLog(pod, 25)
+	//      if err != nil {
+	//          c.Error(fmt.Sprintf("Failed to get logs for pod %s", pod.ObjectMeta.Name), "err", err)
+	//      }
+	//      c.sendErrChan(fmt.Sprintf("Pod %s container terminated. Reason: %s, Exit code: %v\n %s",
+	//          pod.ObjectMeta.Name,
+	//          pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Reason,
+	//          pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode,
+	//          logs))
+	//      c.Info(fmt.Sprintf("Pod %s container terminated. Reason: %s, Exit code: %v",
+	//          pod.ObjectMeta.Name,
+	//          pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Reason,
+	//          pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode))
+	//  }
+	// }
+	return nil
+}
+
+// processNode is called whenever an informer encounters a changed node.
+func (c *Controller) processNode(node *corev1.Node) error {
+	// Keep a running total of the total allocatable resources. Adds information
+	// to a map[nodeName]resourcelist.
+	c.Debug("adding node to resource map", "node", node.ObjectMeta.Name, "resources", node.Status.Allocatable)
+	c.Debug("obtaining resourcemutex Lock()")
+	c.nodeResourceMutex.Lock()
+	c.Debug("obtained resourcemutex Lock()")
+	// Set the allocatable resource amount
+	c.nodeResources[nodeName(node.ObjectMeta.Name)] = node.Status.Allocatable
+	c.nodeResourceMutex.Unlock()
+	c.Debug("returned resourcemutex Lock()")
+	return nil
+}
+
+// Nice unified error handling
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.workqueue.Forget(key)
+		return
+	}
+
+	if c.workqueue.NumRequeues(key) < maxRetries {
+		c.Error("error processing key", "key", key, "error", err.Error())
+		c.sendErrChan(fmt.Sprintf("Error processing key %v: %v", key, err.Error()))
+		c.workqueue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	c.Error("dropping key out of queue", "key", key, "error", err.Error())
+	c.sendErrChan(fmt.Sprintf("Dropping key %q out of queue %v", key, err.Error()))
+	c.workqueue.Forget(key)
+}
+
+// Send a string of the provided error to the callback channel
+func (c *Controller) sendErrChan(err string) {
+	c.Debug("sendErrChan called", "error", err)
+	go func() {
+		c.opts.CbChan <- err
+	}()
+}
+
+// Send a *cloud.Server to wr that's gone bad.
+func (c *Controller) sendBadServer(server *cloud.Server) {
+	c.Debug("sendBadServer called")
+	go func() {
+		c.opts.BadCbChan <- server
+	}()
+}
+
+func (c *Controller) runReqCheck() {
+	c.Debug("runReqCheck() called")
+	// If nodeResources is not initialised, wait.
+	if c.nodeResourcesLen == nil || *c.nodeResourcesLen == 0 {
+		c.Debug("node resources not initialised. Waiting")
+		c.Debug("obtaining RLock()")
+		c.nodeResourceMutex.RLock()
+		c.Debug("obtained RLock()")
+		c.nodeResourcesLen = func(i int) *int { return &i }(len(c.nodeResources))
```

Comment:
This doesn't seem safe. You're protecting nodeResources with a read lock, but not nodeResourcesLen? If runReqCheck() can be called in parallel, odd things could happen. You should get a full lock (on the same or a different mutex) to protect nodeResourcesLen before the if.

Reply:
IIRC runReqCheck() should be called each second indefinitely (So not in parallel). Still worth changing it anyway?